### PR TITLE
Enable release image building

### PR DESCRIPTION
**Pull request by yzhang-23**
**PR Type**

- [ ] Breaking change
- [X] Feature
- [ ] Patch

**Brief Description**
Enable installating of a repo after its successful building, then build a docker image for the installed library and push it to the nwchemex-project registry.

**Not In Scope**

**PR Checklist**

- [ ] [Is documented](https://nwchemex-project.github.io/.github/documenting/index.html)
- [ ] [Is tested](https://nwchemex-project.github.io/.github/testing/index.html)
- [ ] [Adheres to applicable organization standards](https://nwchemex-project.github.io/.github/conventions/index.html)

**TODOs**

- [X] Write the action files.
- [X] Test the actions on my forked branches.

### CODE REVIEW

File: actions/image-build/build.sh [hunk 1]
```diff
@@ -54,12 +54,15 @@ else
 fi
 
 #Step 2: Configure
-
+export INSTALL_PATH=install
```
Comment (keceli) [inline actions/image-build/build.sh:57]: Is this consistent with `CMAKE_PREFIX_PATH` and is there any reason not to put this path in the toolchain file?

**Review by ryanmrichard — APPROVED** [review]

File: .github/workflows/image-c-cpp_tmpl.yaml [hunk 2]
```diff
@@ -57,3 +67,24 @@ jobs:
...
+          password: ${{ secrets.CONTAINER_REPO_TOKEN }}
+      - name: Extract metadata (tags, labels) for Docker
+        if: ${{inputs.gcc-build == true}}
+        id: meta
+        uses: docker/metadata-action@9ec57ed1fcdbf14dcef7dfbe97b2010124a938b7
+        with:
+          images: ghcr.io/nwchemex-project/release_${{needs.Repo_name_lowcase.outputs.lowcase_repo_name}}
+      - name: Build the image and push
+        if: ${{inputs.gcc-build == true}}
+        uses: docker/build-push-action@v4
```
Comment (ryanmrichard) [inline .github/workflows/image-c-cpp_tmpl.yaml:85]: Isn't this going to push every time this workflow is run? Isn't this also the workflow used for testing the PRs? We don't want those to be the same workflow as the deployment is different than the testing.

**Review by ryanmrichard — CHANGES_REQUESTED** [review]
On second thought, this is going to be a huge change to our CI. We should probably setup a dummy organization and iron out all the kinks in an end-to-end solution before pushing them to the NWX org.

File: actions/image-build/build.sh (hunk 1 above)
Comment (yzhang-23) [reply actions/image-build/build.sh:57]: > Is this consistent with `CMAKE_PREFIX_PATH` and is there any reason not to put this path in the toolchain file?

Thanks for pointing out this! I did not realize that the installation directory has been set in the toolchain file generated by build.sh. I can use `CMAKE_PREFIX_PATH` instead of `INSTALL_PATH` here.

File: .github/workflows/image-c-cpp_tmpl.yaml (hunk 2 above)
Comment (yzhang-23) [reply .github/workflows/image-c-cpp_tmpl.yaml:85]: > Isn't this going to push every time this workflow is run? Isn't this also the workflow used for testing the PRs? We don't want those to be the same workflow as the deployment is different than the testing.

So you mean a separated workflow simply to deploy the code, right? What would be the event to trigger this workflow, say a successful testing?

**Comment by yzhang-23** [comment]
> On second thought, this is going to be a huge change to our CI. We should probably setup a dummy organization and iron out all the kinks in an end-to-end solution before pushing them to the NWX org.

I agree. How to copy the entire organization to a new one?

**Comment by ryanmrichard** [comment]
> > On second thought, this is going to be a huge change to our CI. We should probably setup a dummy organization and iron out all the kinks in an end-to-end solution before pushing them to the NWX org.
> 
> I agree. How to copy the entire organization to a new one?

I don't think you need to copy the whole org, just .github, ParallelZone, and PluginPlay. Once you have those going that should be 99% of the battle.

**Comment by yzhang-23** [comment]
> 



> > > On second thought, this is going to be a huge change to our CI. We should probably setup a dummy organization and iron out all the kinks in an end-to-end solution before pushing them to the NWX org.
> > 
> > 
> > I agree. How to copy the entire organization to a new one?
> 
> I don't think you need to copy the whole org, just .github, ParallelZone, and PluginPlay. Once you have those going that should be 99% of the battle.

Ok, I think I have already created an organization for CI testing, just trying to put my forked branches of .github, ParallelZone and PluginPlay into it.

**Comment by ryanmrichard** [comment]
> > 
> 
> > > > On second thought, this is going to be a huge change to our CI. We should probably setup a dummy organization and iron out all the kinks in an end-to-end solution before pushing them to the NWX org.
> > > 
> > > 
> > > I agree. How to copy the entire organization to a new one?
> > 
> > 
> > I don't think you need to copy the whole org, just .github, ParallelZone, and PluginPlay. Once you have those going that should be 99% of the battle.
> 
> Ok, I think I have already created an organization for CI testing, just trying to put my forked branches of .github, ParallelZone and PluginPlay into it.

You should be able to fork the repos as the organization you're trying to put them in.

**Comment by yzhang-23** [comment]
> > > > > On second thought, this is going to be a huge change to our CI. We should probably setup a dummy organization and iron out all the kinks in an end-to-end solution before pushing them to the NWX org.
> > > > 
> > > > 
> > > > I agree. How to copy the entire organization to a new one?
> > > 
> > > 
> > > I don't think you need to copy the whole org, just .github, ParallelZone, and PluginPlay. Once you have those going that should be 99% of the battle.
> > 
> > 
> > Ok, I think I have already created an organization for CI testing, just trying to put my forked branches of .github, ParallelZone and PluginPlay into it.
> 
> You should be able to fork the repos as the organization you're trying to put them in.

Got it. Now all 3 repos are in my CI-test organization.

File: actions/image-build/build.sh (hunk 1 above)
Comment (yzhang-23) [reply actions/image-build/build.sh:57]: > Is this consistent with `CMAKE_PREFIX_PATH` and is there any reason not to put this path in the toolchain file?

After some reconsideration, I think I want to continue to use INSTALL_PATH instead of CMAKE_PREFIX_PATH in the toolchain file. I have two reasons: 
1. CMAKE_PREFIX_PATH may contain other paths (the line `echo "set(CMAKE_PREFIX_PATH $(pwd)/install)"` will definitely be modified in the future), e. g., the path of the installation directory of libint2;
2. Not sure if I put the definition of INSTALL_PATH in the toolchain file, the command line `${cmake_command} -GNinja -H. -Bbuild -DCMAKE_TOOLCHAIN_FILE="${toolchain_file}" -DCMAKE_INSTALL_PREFIX=${INSTALL_PATH}` would recognize `${INSTALL_PATH}` or not.

File: .github/workflows/image-c-cpp_tmpl.yaml (hunk 2 above)
Comment (yzhang-23) [reply .github/workflows/image-c-cpp_tmpl.yaml:94]: > Isn't this going to push every time this workflow is run? Isn't this also the workflow used for testing the PRs? We don't want those to be the same workflow as the deployment is different than the testing.

Added options to install the package, build and push the release docker image only when necessary. Different calls would come from different workflow in the corresponding repo invoked by different events, e. g., PR or push to master.

**Review by keceli — APPROVED** [review]

**Review by ryanmrichard — COMMENTED** [review]
Being blunt, I feel like this PR is part of a prototyping effort, which if that's the case I don't want to merge it. Keep in mind our CI/CD pipeline affects the entire team and we don't want to be using it for prototyping. PRs targeting CI/CD should be moving us towards a sustainable solution, and not contributing to flux. We've admittedly not always been the best at this, but now that we've actually got some software releases I think we need to do better. My hope is that we can do one last overhaul and achieve a stable CI/CD pipeline.

To that end, based on my recollection of past discussions with the CI team, I put together #105 to try and articulate the complexities our CI/CD solution needs to address. I have a design in my head which I think will address them all; however, it relies on a more fine-grained approach to workflows and images than where this seems to be heading. Of course, just because you're not using my design, doesn't mean it won't work, but if we're going to go with a different design there needs to be at least a couple people that understand that design in order to review PRs associated with it. 

Long story short, I'm fine with whatever solution we put together that will address the issues in #105, but until we have that solution, or a path to that solution, I think we're better off not merging anything. 
So what I propose:
- Fully sketch out a CI/CD pipeline design, or
- Get CI/CD working for a copy of the NWChemEx-Project organization and then port CI/CD form the copy to the real project.

**Comment by yzhang-23** [comment]
> 

Ok, I see. Let me make the image-based CI workflow working for PluginPlay and ParallelZone in my own CI-testing organization, then think about merging. I will try to follow the design in #105. Now not all the issues in #105 have been addressed in my planned solution, but I will refine it. Thanks!

**Comment by ryanmrichard** [comment]
> > 
> 
> Ok, I see. Let me make the image-based CI workflow working for PluginPlay and ParallelZone in my own CI-testing organization, then think about merging. I will try to follow the design in #105. Now not all the issues in #105 have been addressed in my planned solution, but I will refine it. Thanks!

I think that's a start, but there's more complexity in play when you get further along. Notably the dependency diamond problem (package A is used by packages B and C; package D then depends on packages B and C and thus depends on A through both B and C).
